File: javascript/src/dns.js
/**
 * DNS TXT cross-verification for SchemaPin v1.4.
 *
 * A tool provider MAY publish a TXT record at `_schemapin.{domain}` containing
 * the public-key fingerprint advertised in `.well-known/schemapin.json`. When
 * present, clients use it as a *second-channel* verification: the DNS
 * credential chain is independent of the HTTPS hosting credential chain, so
 * compromising one does not compromise the other.
 *
 * ## TXT record format
 *
 * ```
 * _schemapin.example.com. IN TXT "v=schemapin1; kid=acme-2026-01; fp=sha256:a1b2c3..."
 * ```
 *
 * Fields:
 * - `v` - version tag (`schemapin1`); required
 * - `fp` - key fingerprint (`sha256:<hex>`); required, lowercase hex
 * - `kid` - optional key id, used for disambiguating multi-key endpoints
 *
 * ## Verification semantics
 *
 * - **Absent record** -> no effect (DNS TXT is optional)
 * - **Present and matching** -> confidence boost
 * - **Present and mismatching** -> hard failure with `DOMAIN_MISMATCH`
 *
 * Use {@link parseTxtRecord} to parse a raw TXT string and {@link verifyDnsMatch}
 * to cross-check it against a discovery document. {@link fetchDnsTxt} performs
 * the actual DNS lookup using Node's built-in `node:dns/promises`.
 */

import { resolveTxt } from 'node:dns/promises';
import { KeyManager } from './crypto.js';

/**
 * Parsed `_schemapin.{domain}` TXT record.
 *
 * @typedef {Object} DnsTxtRecord
 * @property {string} version
 * @property {string|null} kid
 * @property {string} fingerprint - Lowercase fingerprint string, including
 *   the `sha256:` prefix.
 */

/**
 * Parse a raw TXT record value
 * (e.g. `"v=schemapin1; kid=acme-2026-01; fp=sha256:..."`).
 *
 * Whitespace around `;` and `=` is tolerated. Field order is not significant.
 * Unknown fields are ignored (forward-compat). Throws on malformed input.
 *
 * @param {string} value - Raw TXT value
 * @returns {DnsTxtRecord} Parsed record
 * @throws {Error} If required fields are missing or version is unsupported
 */
export function parseTxtRecord(value) {
    let version = null;
    let kid = null;
    let fingerprint = null;

    for (const rawPart of value.split(';')) {
        const part = rawPart.trim();
        if (part.length === 0) {
            continue;
        }
        const eqIdx = part.indexOf('=');
        if (eqIdx === -1) {
            throw new Error(`DNS TXT field missing '=': ${part}`);
        }
        const k = part.slice(0, eqIdx).trim().toLowerCase();
        const v = part.slice(eqIdx + 1).trim();
        if (k === 'v') {
            version = v;
        } else if (k === 'kid') {
            kid = v;
        } else if (k === 'fp') {
            fingerprint = v.toLowerCase();
        }
        // Forward-compat: ignore unknown fields rather than reject.
    }

    if (version === null) {
        throw new Error('DNS TXT record missing required \'v\' field');
    }
    if (version !== 'schemapin1') {
        throw new Error(`DNS TXT unsupported version: ${version}`);
    }
    if (fingerprint === null) {
        throw new Error('DNS TXT record missing required \'fp\' field');
    }
    if (!fingerprint.startsWith('sha256:')) {
        throw new Error(`DNS TXT 'fp' must be sha256:<hex>: ${fingerprint}`);
    }

    return { version, kid, fingerprint };
}

/**
 * Cross-check the DNS TXT record's fingerprint against the discovery document.
 *
 * Computes the SHA-256 fingerprint of `discovery.public_key_pem` and compares
 * it (case-insensitively) against `txt.fingerprint`. Throws on mismatch.
 *
 * @param {Object} discovery - Well-known discovery document
 * @param {DnsTxtRecord} txt - Parsed TXT record
 * @throws {Error} If the fingerprints do not match
 */
export function verifyDnsMatch(discovery, txt) {
    if (!discovery || !discovery.public_key_pem) {
        throw new Error('DNS TXT verify: discovery missing public_key_pem');
    }
    const computed = KeyManager.calculateKeyFingerprint(discovery.public_key_pem).toLowerCase();
    if (computed !== txt.fingerprint) {
        throw new Error(
            `DNS TXT fingerprint mismatch: discovery=${computed}, dns=${txt.fingerprint}`
        );
    }
}

/**
 * Construct the DNS lookup name for a given tool domain.
 *
 * @param {string} domain - Tool domain (e.g. `example.com` or `example.com.`)
 * @returns {string} DNS name (e.g. `_schemapin.example.com`)
 */
export function txtRecordName(domain) {
    return `_schemapin.${domain.replace(/\.$/, '')}`;
}

/**
 * Fetch and parse the `_schemapin.{domain}` TXT record.
 *
 * Returns:
 * - `{ version, kid, fingerprint }` - record present and parseable
 * - `null` - no `_schemapin` TXT record exists for the domain
 *
 * Throws on other resolution errors. Multiple matching TXT chunks are joined
 * per RFC 1464 (concatenation in emit order). Multiple separate TXT records
 * at the same name are not supported - the first valid `v=schemapin1` record
 * wins.
 *
 * @param {string} domain - Tool domain
 * @returns {Promise<DnsTxtRecord|null>}
 */
export async function fetchDnsTxt(domain) {
    const name = txtRecordName(domain);
    let records;
    try {
        records = await resolveTxt(name);
    } catch (err) {
        // `resolveTxt` throws an error with `code` = 'ENOTFOUND' / 'ENODATA'
        // when no TXT record exists. Treat both as "absent".
        if (err && (err.code === 'ENOTFOUND' || err.code === 'ENODATA')) {
            return null;
        }
        throw new Error(`DNS TXT lookup failed for ${name}: ${err.message ?? err}`);
    }

    for (const chunks of records) {
        // RFC 1464: a single TXT record may be split across multiple
        // <character-strings>; node returns these as inner string arrays.
        const joined = chunks.join('');
        if (joined.includes('v=schemapin1')) {
            return parseTxtRecord(joined);
        }
    }
    return null;
}

File: rust/src/dns.rs
//! DNS TXT cross-verification (v1.4).
//!
//! A tool provider MAY publish a TXT record at `_schemapin.{domain}` containing
//! the public-key fingerprint advertised in `.well-known/schemapin.json`. When
//! present, clients use it as a *second-channel* verification: the DNS
//! credential chain is independent of the HTTPS hosting credential chain, so
//! compromising one doesn't compromise the other.
//!
//! ## TXT record format
//!
//! ```text
//! _schemapin.example.com. IN TXT "v=schemapin1; kid=acme-2026-01; fp=sha256:a1b2c3..."
//! ```
//!
//! Fields:
//! - `v` — version tag (`schemapin1`); required
//! - `fp` — key fingerprint (`sha256:<hex>`); required, lowercase hex
//! - `kid` — optional key id, used for disambiguating multi-key endpoints
//!
//! ## Verification semantics
//!
//! - **Absent record** → no effect (DNS TXT is optional)
//! - **Present and matching** → confidence boost (recorded in result warnings is *not* used; absence of mismatch is the signal)
//! - **Present and mismatching** → hard failure with [`crate::error::ErrorCode::DomainMismatch`]
//!
//! Use [`parse_txt_record`] to parse a raw TXT string and [`verify_dns_match`]
//! to cross-check it against a discovery document. With the `dns` feature
//! enabled, [`fetch_dns_txt`] performs the actual DNS lookup.

use crate::crypto;
use crate::error::{Error, ErrorCode};
use crate::types::discovery::WellKnownResponse;

/// Parsed `_schemapin.{domain}` TXT record.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DnsTxtRecord {
    pub version: String,
    pub kid: Option<String>,
    /// Lowercase fingerprint string, including the `sha256:` prefix.
    pub fingerprint: String,
}

/// Parse a raw TXT record value (e.g. `"v=schemapin1; kid=acme-2026-01; fp=sha256:..."`).
///
/// Whitespace around `;` and `=` is tolerated. Field order is not significant.
/// Returns an error if the record is missing the required `v` or `fp` fields,
/// or if the version isn't `schemapin1`.
pub fn parse_txt_record(value: &str) -> Result<DnsTxtRecord, Error> {
    let mut version: Option<String> = None;
    let mut kid: Option<String> = None;
    let mut fp: Option<String> = None;

    for raw_part in value.split(';') {
        let part = raw_part.trim();
        if part.is_empty() {
            continue;
        }
        let (k, v) = part.split_once('=').ok_or_else(|| Error::Verification {
            code: ErrorCode::DiscoveryInvalid,
            message: format!("DNS TXT field missing '=': {}", part),
        })?;
        let k = k.trim().to_ascii_lowercase();
        let v = v.trim();
        match k.as_str() {
            "v" => version = Some(v.to_string()),
            "kid" => kid = Some(v.to_string()),
            "fp" => fp = Some(v.to_ascii_lowercase()),
            // Forward-compat: ignore unknown fields rather than reject.
            _ => {}
        }
    }

    let version = version.ok_or_else(|| Error::Verification {
        code: ErrorCode::DiscoveryInvalid,
        message: "DNS TXT record missing required 'v' field".to_string(),
    })?;
    if version != "schemapin1" {
        return Err(Error::Verification {
            code: ErrorCode::DiscoveryInvalid,
            message: format!("DNS TXT unsupported version: {}", version),
        });
    }
    let fingerprint = fp.ok_or_else(|| Error::Verification {
        code: ErrorCode::DiscoveryInvalid,
        message: "DNS TXT record missing required 'fp' field".to_string(),
    })?;
    if !fingerprint.starts_with("sha256:") {
        return Err(Error::Verification {
            code: ErrorCode::DiscoveryInvalid,
            message: format!("DNS TXT 'fp' must be sha256:<hex>: {}", fingerprint),
        });
    }

    Ok(DnsTxtRecord {
        version,
        kid,
        fingerprint,
    })
}

/// Cross-check the DNS TXT record's fingerprint against the discovery document.
///
/// Returns `Ok(())` when the fingerprint matches the SHA-256 fingerprint of
/// the public key in `discovery.public_key_pem`. Returns
/// [`ErrorCode::DomainMismatch`] otherwise.
pub fn verify_dns_match(discovery: &WellKnownResponse, txt: &DnsTxtRecord) -> Result<(), Error> {
    let computed = crypto::calculate_key_id(&discovery.public_key_pem)?.to_ascii_lowercase();
    if computed == txt.fingerprint {
        Ok(())
    } else {
        Err(Error::Verification {
            code: ErrorCode::DomainMismatch,
            message: format!(
                "DNS TXT fingerprint mismatch: discovery={}, dns={}",
                computed, txt.fingerprint
            ),
        })
    }
}

/// Construct the DNS lookup name for a given tool domain.
pub fn txt_record_name(domain: &str) -> String {
    format!("_schemapin.{}", domain.trim_end_matches('.'))
}

/// Fetch and parse the `_schemapin.{domain}` TXT record. Behind the `dns` feature.
///
/// Returns:
/// - `Ok(Some(record))` — record present and parseable
/// - `Ok(None)` — no `_schemapin` TXT record exists for the domain
/// - `Err(_)` — DNS resolution error or the record exists but is malformed
///
/// Multiple matching TXT chunks are joined per RFC 1464 (concatenation in
/// emit order). Multiple separate TXT records at the same name are not
/// supported — the first valid record wins.
#[cfg(feature = "dns")]
pub async fn fetch_dns_txt(domain: &str) -> Result<Option<DnsTxtRecord>, Error> {
    use hickory_resolver::error::ResolveErrorKind;
    use hickory_resolver::TokioAsyncResolver;

    let name = txt_record_name(domain);
    let resolver = TokioAsyncResolver::tokio(Default::default(), Default::default());
    let lookup = match resolver.txt_lookup(&name).await {
        Ok(l) => l,
        Err(e) => {
            if matches!(e.kind(), ResolveErrorKind::NoRecordsFound { .. }) {
                return Ok(None);
            }
            return Err(Error::Verification {
                code: ErrorCode::DiscoveryFetchFailed,
                message: format!("DNS TXT lookup failed for {}: {}", name, e),
            });
        }
    };

    for record in lookup.iter() {
        // hickory yields TxtData as Vec<Box<[u8]>>; concatenate chunks per RFC 1464.
        let joined: String = record
            .iter()
            .map(|chunk| String::from_utf8_lossy(chunk).into_owned())
            .collect::<Vec<_>>()
            .join("");
        if joined.contains("v=schemapin1") {
            return parse_txt_record(&joined).map(Some);
        }
    }
    Ok(None)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::crypto::generate_key_pair;
    use crate::discovery::build_well_known_response;

    #[test]
    fn test_parse_full_record() {
        let r = parse_txt_record("v=schemapin1; kid=acme-2026-01; fp=sha256:abcd1234").unwrap();
        assert_eq!(r.version, "schemapin1");
        assert_eq!(r.kid.as_deref(), Some("acme-2026-01"));
        assert_eq!(r.fingerprint, "sha256:abcd1234");
    }

    #[test]
    fn test_parse_minimal_record() {
        let r = parse_txt_record("v=schemapin1;fp=sha256:abc").unwrap();
        assert_eq!(r.version, "schemapin1");
        assert_eq!(r.kid, None);
        assert_eq!(r.fingerprint, "sha256:abc");
    }

    #[test]
    fn test_parse_lowercases_fingerprint() {
        let r = parse_txt_record("v=schemapin1; fp=SHA256:ABCDEF").unwrap();
        assert_eq!(r.fingerprint, "sha256:abcdef");
    }

    #[test]
    fn test_parse_tolerates_whitespace_and_order() {
        let r = parse_txt_record("  fp = sha256:beef ;  v = schemapin1  ").unwrap();
        assert_eq!(r.version, "schemapin1");
        assert_eq!(r.fingerprint, "sha256:beef");
    }

    #[test]
    fn test_parse_ignores_unknown_fields() {
        let r = parse_txt_record("v=schemapin1; fp=sha256:abc; future=ignoreme").unwrap();
        assert_eq!(r.fingerprint, "sha256:abc");
    }

    #[test]
    fn test_parse_missing_v_fails() {
        let e = parse_txt_record("fp=sha256:abc").unwrap_err();
        assert!(matches!(
            e,
            Error::Verification {
                code: ErrorCode::DiscoveryInvalid,
                ..
            }
        ));
    }

    #[test]
    fn test_parse_missing_fp_fails() {
        let e = parse_txt_record("v=schemapin1").unwrap_err();
        assert!(matches!(
            e,
            Error::Verification {
                code: ErrorCode::DiscoveryInvalid,
                ..
            }
        ));
    }

    #[test]
    fn test_parse_unsupported_version_fails() {
        assert!(parse_txt_record("v=schemapin99; fp=sha256:abc").is_err());
    }

    #[test]
    fn test_parse_fp_without_sha256_prefix_fails() {
        assert!(parse_txt_record("v=schemapin1; fp=abc").is_err());
    }

    #[test]
    fn test_parse_field_without_equals_fails() {
        assert!(parse_txt_record("v=schemapin1; broken").is_err());
    }

    #[test]
    fn test_verify_match() {
        let kp = generate_key_pair().unwrap();
        let fp = crypto::calculate_key_id(&kp.public_key_pem).unwrap();
        let discovery = build_well_known_response(&kp.public_key_pem, None, vec![], "1.4");
        let txt = DnsTxtRecord {
            version: "schemapin1".to_string(),
            kid: None,
            fingerprint: fp.to_ascii_lowercase(),
        };
        verify_dns_match(&discovery, &txt).unwrap();
    }

    #[test]
    fn test_verify_mismatch_returns_domain_mismatch() {
        let kp = generate_key_pair().unwrap();
        let discovery = build_well_known_response(&kp.public_key_pem, None, vec![], "1.4");
        let txt = DnsTxtRecord {
            version: "schemapin1".to_string(),
            kid: None,
            fingerprint: "sha256:0000000000000000000000000000000000000000000000000000000000000000"
                .to_string(),
        };
        let err = verify_dns_match(&discovery, &txt).unwrap_err();
        match err {
            Error::Verification { code, .. } => assert_eq!(code, ErrorCode::DomainMismatch),
            other => panic!("unexpected error: {:?}", other),
        }
    }

    #[test]
    fn test_txt_record_name_strips_trailing_dot() {
        assert_eq!(txt_record_name("example.com"), "_schemapin.example.com");
        assert_eq!(txt_record_name("example.com."), "_schemapin.example.com");
    }
}
